src/dparse/strings.d

```diff
@@ -0,0 +1,694 @@
+{
+	// there are no 1 character strings
+	if (literal.length < 2)
+		return false;
+
+	// check for valid start
+	bool allowPostfix;
+	switch (literal[0])
+	{
+	case 'r': // WysiwygString
+	case 'x': // HexString
+		if (literal[1] != '"')
+			return false;
+		stringCloseChar = '"';
+		allowPostfix = true;
+		prefixLength = 2;
+		break;
+	case 'q': // DelimitedString
+		if (literal[1] == '{')
+			stringCloseChar = '}';
+		else if (literal[1] == '"')
+			stringCloseChar = '"';
+		else
+			return false;
+
+		allowPostfix = false;
+		prefixLength = 2;
+		break;
+	case '`':
+	case '"':
+		stringCloseChar = literal[0];
+		allowPostfix = true;
+		parseEscapes = stringCloseChar == '"';
+		prefixLength = 1;
+		break;
+	default:
+		return false;
+	}
+
+	if (allowPostfix && literal[$ - 1].among!('c', 'w', 'd'))
+	{
+		hasPostfix = true;
+		literal = literal[0 .. $ - 1];
+	}
+
+	if (literal.length <= prefixLength || literal[$ - 1] != stringCloseChar)
+		return false;
+
+	if (parseEscapes)
+	{
+		// check if end escapes the quote, making this an invalid string
+		auto end = literal[0 .. $ - 1].lastIndexOfNeither("\\");
+		if (end != -1)
+		{
+			// don't need to subtract 1
+			size_t countBackslashes = literal.length - end;
+
+			if ((countBackslashes % 2) != 0)
+				return false; // uneven backslash count -> invalid end
+		}
+	}
+
+	return true;
+}
+
+/// ditto
+bool isStringLiteral(const(char)[] literal)
+{
+	char stringCloseChar;
+	bool hasPostfix, parseEscapes;
+	int prefixLength;
+	return isStringLiteral(literal, stringCloseChar, hasPostfix, parseEscapes,
+		prefixLength);
+}
+
+///
+unittest
+{
+	assert(isStringLiteral(`"hello"`));
+	assert(isStringLiteral(`"hello world!"`));
+	assert(isStringLiteral(`r"hello world!"c`));
+	assert(isStringLiteral(`r"hello world!"d`));
+	assert(isStringLiteral(`q{cool}`));
+	assert(isStringLiteral(`q{cool\}`));
+	assert(isStringLiteral(`"\\"`));
+	assert(!isStringLiteral(`"\\\"`));
+	assert(isStringLiteral(`"\\\\"`));
+	assert(isStringLiteral(`"a\\\\"`));
+	assert(isStringLiteral(`""`));
+	assert(isStringLiteral(`q""`));
+	assert(isStringLiteral(`x""`));
+	assert(!isStringLiteral(``));
+	assert(!isStringLiteral(`"`));
+	assert(!isStringLiteral(`w""`));
+	assert(!isStringLiteral(`hello"`));
+	assert(!isStringLiteral(`"hello`));
+	assert(!isStringLiteral(`"hello world`));
+	assert(!isStringLiteral(`hello world`));
+	assert(!isStringLiteral(`r"`));
+	assert(!isStringLiteral(`rr"ok"`));
+	assert(!isStringLiteral(`x"`));
+	assert(!isStringLiteral(`x" `));
+	assert(!isStringLiteral(`qqqq`));
+}
+
+/// Defines different handler types what to do when invalid escape sequences are
+/// found inside $(LREF unescapeString).
+enum InvalidEscapeAction
+{
+	/// keep the backslash character as well as the escape characters in the
+	/// string like in the input string.
+	keep = 0,
+	/// Ignore and skip offending characters, drop them from the output. Named
+	/// character entities are still being included like keep as they are not
```

Comment:
I think that it would be more clear to write this as "included like \`keep\` as they" to emphasize that the other enum member is being referenced.